src/passes/J2clOpts.cpp

```diff
@@ -0,0 +1,210 @@
+#include "passes.h"
+#include "wasm.h"
+
+namespace wasm {
+
+namespace {
+
+static bool isOnceFunction(Function* f) {
+  return f->name.hasSubstring("_@once@_");
+}
+
+using AssignmentCountMap = std::unordered_map<Name, Index>;
+
+// TODO: parallel
+class GlobalAssignmentCollector
+  : public WalkerPass<PostWalker<GlobalAssignmentCollector>> {
+public:
+  GlobalAssignmentCollector(AssignmentCountMap& assignmentCounts)
+    : assignmentCounts(assignmentCounts) {}
+
+  void visitGlobalSet(GlobalSet* curr) {
+    // Avoid optimizing class initialization condition variable itself.
+    if (curr->name.startsWith("f_$initialized__")) {
+      return;
+    }
+    assignmentCounts[curr->name]++;
+  }
+
+private:
+  AssignmentCountMap& assignmentCounts;
+};
+
+// TODO: parallel
+class ConstantHoister : public WalkerPass<PostWalker<ConstantHoister>> {
+public:
+  ConstantHoister(AssignmentCountMap& assignmentCounts)
+    : assignmentCounts(assignmentCounts) {}
+
+  int optimized = 0;
+
+  void visitFunction(Function* curr) {
+    if (!isOnceFunction(curr)) {
+      return;
+    }
+    int optimizedBefore = optimized;
+    if (auto* block = curr->body->dynCast<Block>()) {
+      for (auto*& expr : block->list) {
+        maybeHoistConstant(expr);
+      }
+    } else {
+      maybeHoistConstant(curr->body);
+    }
+
+    if (optimized != optimizedBefore) {
+      // We introduced "nop" instruction. Run the vacuum to cleanup.
+      // TODO: maybe we should not introduce "nop" in the first place and try
+      // removing instead.
+      PassRunner runner(getModule());
+      runner.add("vacuum");
+      runner.setIsNested(true);
+      runner.runOnFunction(curr);
+    }
+  }
+
+private:
+  void maybeHoistConstant(Expression* expr) {
+    auto set = expr->dynCast<GlobalSet>();
+    if (!set) {
+      return;
+    }
+
+    if (assignmentCounts[set->name] != 1) {
+      // The global assigned in multiple places, so it is not safe to
+      // hoist them as global constants.
+      return;
+    }
+
+    if (!GlobalUtils::canInitializeGlobal(*getModule(), set->value)) {
+      // It is not a valid constant expression so cannot be hoisted to
+      // global init.
+      return;
+    }
+
+    // Move it to global init and mark it as immutable.
+    auto global = getModule()->getGlobal(set->name);
+    global->init = set->value;
+    global->mutable_ = false;
+    ExpressionManipulator::nop(expr);
+
+    optimized++;
+  }
+
+  AssignmentCountMap& assignmentCounts;
+};
+
+using RenamedFunctions = std::map<Name, Name>;
+
+// TODO: parallel
+class RenameFunctions : public WalkerPass<PostWalker<RenameFunctions>> {
+public:
+  RenameFunctions(RenamedFunctions& renamedFunctions)
+    : renamedFunctions(renamedFunctions) {}
+  void visitFunction(Function* curr) {
+    if (!isOnceFunction(curr)) {
+      return;
+    }
+    auto* body = curr->body;
+    if (Measurer::measure(body) <= 2) {
+      // Once function has become trivial: rename so could be removed/inlined.
```

Comment:
This could be part of a comment on the class itself. In general it seems like renaming is not the main point here, though, so perhaps this class could be described as "marks trivial once functions for removal"; the mechanism for doing that is to rename them, but not the point in itself.